### DISCUSSION TITLE
How to run generator after .vsix installation ?

### DISCUSSION THREAD

**Discussion by open-domain** [eclipse-langium/langium#855] [Q&A] [answered]
Hi, I am relatively new to Langium and vscode extension, I am struggling with running the generator for code generation. 

I have packaged(.vsix) my project and installed in vscode. Everything else works except generator. 

I can able to generate code from Langium project using "**node ./bin/cli generate xx.xxx**", how can I generate the code after installing the extension ?

I tried a bit to make use of "Command Palette", but its quite complicated and I don't have confidence if I am going the right way.

Also in later point of time, I have to generate the code through command line without opening the vscode for CI/CD purpose. How can i achieve this ?

Thank you 
Maha

**Comment by msujew** [comment, marked as answer]
Hey @open-domain,

so Langium doesn't really provide a lot of guidance on how you can run the generator, given that there are a lot of different use cases how and when people want to generate their code. The CLI we offer by default is meant to be installed as a separate package, similar to how langium projects rely on the [langium-cli](https://github.com/langium/langium/tree/main/packages/langium-cli) package.

Another way would be to have your generator run on every document change. You could hook your generator into the document lifecycle by calling [`onBuildPhase`](https://github.com/langium/langium/blob/c5d912ba4ad35c2494f4598b1c627ea511782cb3/packages/langium/src/workspace/document-builder.ts#L58). That would then run the generator on every changed document in the editor.

You could also include the generator as a command. One way this could work is by registering a command on the language server side (using the [`AbstractExecuteCommandHandler`](https://github.com/langium/langium/blob/c5d912ba4ad35c2494f4598b1c627ea511782cb3/packages/langium/src/lsp/execute-command-handler.ts#L21) and call the command from the language client whenever the command is triggered in vscode.

> Also in later point of time, I have to generate the code through command line without opening the vscode for CI/CD purpose. How can i achieve this ?

The best way to achieve this is probably the first option which I've outlined, i.e. creating a separate cli package for your language, which users can include in their `devDependencies` to run during CI/CD.

**Reply by open-domain** [reply]
Hallo @msujew 

Thank you for your response. It actually actually helped me. I am having fun with Langium.

I managed to make execute generator through command(3rd solution).

However, "onBuildPhase" solution seems to be relatively easy to realize but it might not fit for me, I am dealing with huge set of model with multiple files and multiple cross references and running generator actually takes some seconds to finish. Do you still think if this solution is still viable ? Do you also think running the generator each time I change the model/document a bad solution ?

Regarding solution of creating separate CLI for generator, I lean more towards this solution. Do you have any sample project for reference ?

Keep up the good work.

MfG
Maha

**Reply by msujew** [reply]
I'm glad to hear you're having a good time with Langium :)

> Do you still think if this solution is still viable ? Do you also think running the generator each time I change the model/document a bad solution ?

The cool thing about `onBuildPhase` is that it only ever passes a subset of all documents, usually those affected by the change. Meaning those that either have changed or are directly related to those via cross references. So only a handful of documents will be actually be generated at any given time. If your language is heavily interdependent or you need to generate the whole project all at once, this however could still result in a large delta. If that's not the case, it could be a good solution for this. If you implement a callback for the method, make sure to regularly allow to interrupt the generator process and listen to the cancellation token using the `interruptAndCheck` function. Otherwise the generator will block the normal document lifecycle.

> Regarding solution of creating separate CLI for generator, I lean more towards this solution. Do you have any sample project for reference ?

The best sample I can give is the [`langium-cli`](https://github.com/langium/langium/tree/main/packages/langium-cli) itself. It's written very similar to how the CLI in the yeoman generated project works. The easiest way to split the CLI code from the DSL code would be to use something like a monorepo structure. We use npm workspaces for that.
